shelley/chain-and-ledger/executable-spec/src/Shelley/Spec/Ledger/Genesis.hs

```diff
@@ -0,0 +1,176 @@
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE OverloadedStrings  #-}
+
+module Shelley.Spec.Ledger.Genesis
+  ( ShelleyGenesisStaking(..)
+  , ShelleyGenesis(..)
+  , emptyGenesisStaking
+  , sgActiveSlotCoeff
+  ) where
+
+import Data.Aeson (FromJSON (..), ToJSON (..), Value (..), (.=), (.:))
+import qualified Data.Aeson as Aeson
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import Data.Time (NominalDiffTime, UTCTime)
+import Data.Word (Word32, Word64)
+import GHC.Generics (Generic)
+
+import Cardano.Crypto (ProtocolMagicId)
+import Cardano.Prelude (Natural, NoUnexpectedThunks)
+import Cardano.Slotting.Slot (EpochSize)
+
+import qualified Shelley.Spec.Ledger.Address as SL
+import Shelley.Spec.Ledger.BaseTypes (ActiveSlotCoeff)
+import qualified Shelley.Spec.Ledger.BaseTypes as SL
+import qualified Shelley.Spec.Ledger.Coin as SL
+import Shelley.Spec.Ledger.Crypto (Crypto)
+import qualified Shelley.Spec.Ledger.Keys as SL
+import qualified Shelley.Spec.Ledger.PParams as SL
+import qualified Shelley.Spec.Ledger.TxData as SL
+
+-- | Genesis Shelley staking configuration.
+--
+-- This allows us to configure some initial stake pools and delegation to them,
+-- in order to test Praos in a static configuration, without requiring on-chain
+-- registration and delegation.
+--
+-- For simplicity, pools defined in the genesis staking do not pay deposits for
+-- their registration.
+data ShelleyGenesisStaking c = ShelleyGenesisStaking {
+      -- | Pools to register
+      --
+      --   The key in this map is the hash of the public key of the _pool_. This
+      --   need not correspond to any payment or staking key, but must correspond
+      --   to the cold key held by 'TPraosIsCoreNode'.
+      sgsPools :: !(Map (SL.KeyHash 'SL.StakePool c) (SL.PoolParams c))
+      -- | Stake-holding key hash credentials and the pools to delegate that stake
+      -- to. We require the raw staking key hash in order to:
+      --
+      -- - Avoid pointer addresses, which would be tricky when there's no slot or
+      --   transaction to point to.
+      -- - Avoid script credentials.
+    , sgsStake :: !(Map (SL.KeyHash 'SL.Staking c) (SL.KeyHash 'SL.StakePool c))
+    }
+  deriving stock    (Eq, Show, Generic)
+  deriving anyclass (NoUnexpectedThunks)
+
+-- | Empty genesis staking
+emptyGenesisStaking :: ShelleyGenesisStaking c
+emptyGenesisStaking = ShelleyGenesisStaking
+  { sgsPools = Map.empty
+  , sgsStake = Map.empty
+  }
+
+-- | Shelley genesis information
+--
+-- Note that this is needed only for a pure Shelley network, hence it being
+-- defined here rather than in its own module. In mainnet, Shelley will
+-- transition naturally from Byron, and thus will never have its own genesis
+-- information.
+data ShelleyGenesis c = ShelleyGenesis {
+      sgSystemStart           :: !UTCTime
+    , sgNetworkMagic          :: !Word32
+    , sgNetworkId             :: !SL.Network
+    , sgProtocolMagicId       :: !ProtocolMagicId
```

Comment:
While we're moving things around, can we please merge these two things or at least name them better.

The terms "network magic" and the "protocol magic" are two terms for the same thing.

The `ProtocolMagicId` is the actual network/protocol magic number. Just a number. This is used in the network handshake protocol, and is used in some of the Byron era crypto, and in some Byron addresses.

The `SL.Network` refers to whether the blockchain is a testnet, or a mainnet, since this info is encoded into Shelley addresses (much like some Byron addresses contain the full magic number, but here just a small bit field with an enumeration of testnet vs mainnet1, mainnet2 etc).

These two are related in the sense that the mainnet always uses a fixed network/protocol magic number.

In the Cardano API we use this type to represent both in one go:
```
data Network
  = Mainnet
  | Testnet !NetworkMagic
```

This means the network magic for mainnet is implicit and has to be hard coded, but that is ok. Alternatively to support multiple mainnets we could use:

```
data Network
  = Mainnet !Int !NetworkMagic
  | Testnet !NetworkMagic
```
That `Int` there refers to which mainnet (Shelley addresses support up to 15), plus the corresponding network magic.